### DISCUSSION TITLE
What data is cached in Atlas, and what are the details.

### DISCUSSION THREAD

**Discussion by benbelow** [Anthony-Nolan/Atlas#787] [Q&A] [answered]
Question from @daria-sorokina-da 

"Caching - what's cached? where? which types of caches (in-memory?)? how to invalidate?"

**Comment by zabeen** [comment, marked as answer]
A package called `LazyCache` is used extensively to provide in-memory caches.

Two types of cache length are configured: [Transient and Persistent](https://github.com/Anthony-Nolan/Atlas/blob/89bc0e94579ee4fe65d115d3f2eeec7df38d08e8/Atlas.Common/Caching/CacheProviders.cs#L6)

 ## [Transient Cache](https://github.com/Anthony-Nolan/Atlas/blob/6b97a3fc28945fef54089a148674076cd9f8cf2f/Atlas.Common/Caching/CacheRegisteringExtensions.cs#L17)

* Counter intuitively this cache has an incredibly long lifetime (~10 years)
* It is "transient" as it will only be registered as a transient dependency in DI - meaning a new one is created whenever we need it
* This cache is used for caching configuration values for which we need the data to be consistent across a request, even if it is not difficult to re-calculate
  * e.g. which matching database is currently active

## [Persistent Cache](https://github.com/Anthony-Nolan/Atlas/blob/6b97a3fc28945fef54089a148674076cd9f8cf2f/Atlas.Common/Caching/CacheRegisteringExtensions.cs#L28)

* This cache persists across search requests, and expires once per day
* It is used for quickly looking up metadata without needing to make a request to a third party store (e.g. Azure Storage)
  * e.g. HLA Metadata dictionary - all tables are cached in memory for quick lookup
  * e.g. MAC dictionary
  * e.g. HF sets in match prediction
* If the data in this cache becomes out of date, a service restart of all affected functions apps is necessary to force cache regeneration
  * see #745 for an update that would force cache invalidation for HMD updates, similar logic could be applied for other cached metadata
